wallet/wallet-controller/src/sync/mod.rs

```diff
@@ -0,0 +1,159 @@
+// Copyright (c) 2023 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::{sync::Arc, time::Duration};
+
+use common::{chain::ChainConfig, primitives::BlockHeight};
+use node_comm::node_traits::NodeInterface;
+use tokio::sync::Mutex;
+use wallet::DefaultWallet;
+
+#[derive(thiserror::Error, Debug)]
+pub enum SyncError {
+    #[error("Unexpected RPC error: {0}")]
+    UnexpectedRpcError(String),
+    #[error("Unexpected wallet error: {0}")]
+    UnexpectedWalletError(wallet::WalletError),
+}
+
+/// Sync the wallet state (blocks) from the node.
+/// Returns true if the wallet state has changed and false otherwise.
+async fn sync_blocks<T: NodeInterface>(
+    chain_config: &ChainConfig,
+    rpc_client: &mut T,
+    wallet: Arc<Mutex<DefaultWallet>>,
+) -> Result<bool, SyncError> {
+    // TODO: Make it more efficient: download blocks concurrently and send fewer requests
+    let wallet_block_height = match wallet
```

Comment:
The height shouldn't be used to determine if there's a change. You should instead store the best block id. It's always possible to switch between two different blockchain states while maintaining the same height.

Reply:
Okay, thanks. This should be fixed now.